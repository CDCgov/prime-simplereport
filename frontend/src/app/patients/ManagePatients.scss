.sr-showing-patients-on-page {
  margin-left: 2rem;
  color: #555;
  font-size: 50%;
  vertical-align: middle;
  font-weight: normal;
}

.sr-patient-list {
  .sr-patient-row th,
  .sr-patient-row td {
    /* override USWDS force-white */
    background: inherit;
  }

  .sr-patient-row-header {
    max-width: 18.75rem;
  }

  .sr-patient-row--removed {
    text-decoration: line-through;
  }

  .rc-menu-button {
    background: inherit;
    padding: 0;
    border: 0;
  }
}

.patient-form-deletion-field {
  width: calc(100% + 2.5rem);
}
