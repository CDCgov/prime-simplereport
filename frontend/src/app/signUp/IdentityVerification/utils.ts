import moment from "moment";
import * as yup from "yup";

import { phoneNumberIsValid } from "../../patients/personSchema";
import { isValidDate } from "../../utils/date";

export const getAnswerKey = (index: number) => `answer${index + 1}`;

// Takes date in format YYYY-MM-DD
export function isValidBirthdate(date: string | undefined) {
  if (date === undefined || date === "") {
    return false;
  }
  if (date.split("-").length === 3 && date.split("-")[0].length < 4) {
    return false;
  }
  if (!isValidDate(date)) {
    return false;
  }
  const parsedDate = moment(date, "YYYY-MM-DD");
  if (parsedDate.year() < 1900) {
    return false;
  }
  if (parsedDate.isAfter(moment())) {
    return false;
  }
  return true;
}

export const toOptions = (
  choices: string[]
): { value: string; label: string }[] =>
  choices.map((choice, index) => {
    return { value: String(index + 1), label: choice };
  });

export const initAnswers = (questionSet: Question[]): Nullable<Answers> =>
  questionSet.reduce((answers, _question, index) => {
    answers[getAnswerKey(index)] = null;
    return answers;
  }, {} as Nullable<Answers>);

export const buildSchema = (questionSet: Question[]): yup.SchemaOf<Answers> =>
  yup.object(
    questionSet.reduce((answers, _question, index) => {
      answers[getAnswerKey(index)] = yup
        .string()
        .nullable()
        .required("This field is required");
      return answers;
    }, {} as { [key: string]: any })
  );

// Put the questions in the same order as received from the server
// Then parse the answers in the same order
export const answersToArray = (answers: Answers): number[] =>
  Object.keys(answers)
    .sort()
    .map((key) => parseInt(answers[key]));

export const initPersonalDetails = (
  orgExternalId: string,
  firstName: string,
  middleName: string,
  lastName: string
): IdentityVerificationRequest => ({
  firstName,
  middleName,
  lastName,
  dateOfBirth: "",
  email: "",
  phoneNumber: "",
  streetAddress1: "",
  city: "",
  state: "",
  zip: "",
  orgExternalId,
});

export const initPersonalDetailsErrors = (): Record<
  keyof IdentityVerificationRequest,
  string
> => ({
  firstName: "",
  middleName: "",
  lastName: "",
  dateOfBirth: "",
  email: "",
  phoneNumber: "",
  streetAddress1: "",
  streetAddress2: "",
  city: "",
  state: "",
  zip: "",
  orgExternalId: "",
});

// this is the regex experian uses for street validation
const experianStreetRegex = new RegExp(
  "^([a-zA-Z0-9# \\-'$ / \\.]{1,60}){1}$",
  "m"
);

// this is the regex experian uses for zip validation
const experianZipRegex = new RegExp("^([\\d]{5}([\\-]?[\\d]{4})?){1}$", "m");

export const personalDetailsSchema: yup.SchemaOf<IdentityVerificationRequest> =
  yup.object().shape({
    firstName: yup.string().required("First name is required"),
    middleName: yup.string().nullable(),
    lastName: yup.string().required("Last name is required"),
    dateOfBirth: yup
      .string()
      .test("birth-date", "A valid date of birth is required", isValidBirthdate)
      .required("A valid date of birth is required"),
    email: yup
      .string()
      .email("A valid email address is required")
      .required("A valid email address is required"),
    phoneNumber: yup
      .mixed()
      .test(
        "phone-number",
        "A valid phone number is required",
        phoneNumberIsValid
      )
      .required(),
    streetAddress1: yup
      .string()
      .matches(experianStreetRegex, "A valid street address is required")
      .required("A valid street address is required"),
    streetAddress2: yup
      .string()
      .nullable()
      .notRequired()
      .matches(experianStreetRegex, {
        message: "Street 2 contains invalid symbols",
        excludeEmptyString: true,
      }),
    city: yup.string().required("City is required"),
    state: yup.string().required("State is required"),
    zip: yup
      .string()
      .matches(experianZipRegex, "A valid ZIP code is required")
      .required("A valid ZIP code is required"),
    orgExternalId: yup.string().required("Organization ID is required"),
  });
