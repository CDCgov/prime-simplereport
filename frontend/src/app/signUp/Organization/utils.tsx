import * as yup from "yup";
import { ReactElement } from "react";

import { TextWithTooltip } from "../../commonComponents/TextWithTooltip";
import { phoneNumberIsValid } from "../../patients/personSchema";
import { liveJurisdictions } from "../../../config/constants";
import Alert from "../../commonComponents/Alert";

import { OrganizationCreateRequest } from "./OrganizationForm";

export const OrganizationTypeEnum: { [key in OrganizationType]: string } = {
  airport: "Airport/Transit station",
  assisted_living: "Assisted living facility",
  camp: "Camp",
  university: "College/University",
  correctional_facility: "Correctional facility",
  employer: "Employer",
  fqhc: "Federally Qualified Health Center (FQHC)",
  government_agency: "Government agency",
  shelter: "Homeless shelter",
  hospice: "Hospice",
  hospital: "Hospital or clinic",
  k12: "K-12 school",
  lab: "Lab",
  nursing_home: "Nursing home",
  other: "Other",
  pharmacy: "Pharmacy",
  primary_care: "Primary care/Mental health outpatient",
  treatment_center: "Substance use disorder treatment center",
  urgent_care: "Urgent care",
};

// This object contains a map of all the fields in the form
// It also contains the input labels and whether the input is required
// The object is iterated over in OrganizationForm.tsx using
// Object.entries() to generate the form
export const organizationFields = [
  [
    "name",
    <TextWithTooltip
      text="Organization name"
      tooltip="Organizations have multiple testing facilities or locations as part of their network."
      position="right"
    />,
    true,
    "",
  ],
  ["state", "Organization state", true, ""],
  ["type", "Organization type", true, ""],
  ["banner", "Organization administrator", false, ""],
  ["firstName", "First name", true, ""],
  ["middleName", "Middle name", false, ""],
  ["lastName", "Last name", true, ""],
  ["email", "Work email", true, "Enter your individual work email address."],
  [
    "workPhoneNumber",
    "Work phone number",
    true,
    "Enter your direct work phone number.",
  ],
].reduce((fields, field) => {
  fields[field[0] as keyof OrganizationCreateRequest] = {
    label: field[1] as string | React.ReactNode,
    required: field[2] as boolean,
    hintText: field[3] as string,
  };
  return fields;
}, {} as { [key: string]: { label: string | React.ReactNode; required: boolean; hintText: string } });

export const initOrg = (): OrganizationCreateRequest => ({
  name: "",
  type: "",
  state: "",
  firstName: "",
  middleName: "",
  lastName: "",
  email: "",
  workPhoneNumber: "",
});

export const initOrgErrors = (): Record<
  keyof OrganizationCreateRequest,
  string
> => ({
  name: "",
  type: "",
  state: "",
  firstName: "",
  middleName: "",
  lastName: "",
  email: "",
  workPhoneNumber: "",
});

const getStateErrorMessage = (param: any) =>
  param.value !== "" ? (
    <>
      SimpleReport isn't available yet in your state. View{" "}
      <a
        href="https://simplereport.gov/using-simplereport/manage-facility-info/find-supported-jurisdictions"
        target="_blank"
        rel="noopener noreferrer"
      >
        supported jurisdictions
      </a>{" "}
      or sign up for our{" "}
      <a
        href="https://simplereport.gov/waitlist"
        target="_blank"
        rel="noopener noreferrer"
      >
        waitlist
      </a>
      .
    </>
  ) : (
    "Organization state is required"
  );

export const organizationSchema: yup.SchemaOf<OrganizationCreateRequest> = yup
  .object()
  .shape({
    name: yup.string().required("Organization name is required"),
    type: yup
      .mixed()
      .oneOf(Object.keys(OrganizationTypeEnum), "Organization type is required")
      .required(),
    state: yup
      .mixed()
      .oneOf(liveJurisdictions, getStateErrorMessage)
      .required(),
    firstName: yup.string().required("First name is required"),
    middleName: yup.string().nullable(),
    lastName: yup.string().required("Last name is required"),
    email: yup
      .string()
      .email("A valid email address is required")
      .required("A valid email address is required"),
    workPhoneNumber: yup
      .mixed()
      .test("", "A valid phone number is required", phoneNumberIsValid)
      .required(),
  });

export const organizationBackendErrors = (error: string): ReactElement => {
  switch (error) {
    // Unusable org name.  Name probably doesn't contain any alphanumeric characters
    case "The organization name is empty.":
    case "The organization name is invalid.":
      return (
        <Alert type="error" title="Invalid organization name">
          The organization name you entered is invalid. Please double check and
          re-enter the organization name.
        </Alert>
      );
    // Duplicate org. Admin user is attempting to resign up but has already completed identity verification.
    case "Duplicate organization with admin user who has completed identity verification.":
      return (
        <Alert type="error" title="Duplicate organization">
          Your organization is already registered with SimpleReport. Check your
          email for instructions on setting up your account.
        </Alert>
      );
    // Duplicate org. Admin user is attempting to resign up and hasn't finished identity verification.
    case "Duplicate organization with admin user that has not completed identity verification.":
      return (
        <Alert type="error" title="Duplicate organization">
          Your organization is already registered with SimpleReport. To begin
          using it, schedule a time to complete our{" "}
          <a href="https://outlook.office365.com/owa/calendar/Helpdesk@HHSGOV.onmicrosoft.com/bookings/s/HOX7KgZruESJksIgC8qVxQ2">
            online identity verification.
          </a>{" "}
        </Alert>
      );
    // Duplicate org. Non-admin user is attempting to reregister the organization.
    case "This organization has already registered with SimpleReport.":
      return (
        <Alert type="error" title="Duplicate organization">
          This organization already has a SimpleReport account. Please contact
          your organization administrator to request access.
        </Alert>
      );
    // Email already exists in SimpleReport (user is potentially already in another org.)
    case "This email address is already associated with a SimpleReport user.":
      return (
        <Alert type="error" title="Email already registered">
          This email address is already registered with SimpleReport. Please
          contact your organization administrator or{" "}
          <a href="mailto:support@simplereport.gov">support@simplereport.gov</a>{" "}
          for help.
        </Alert>
      );
    // Okta internal error.
    case "An unknown error occurred when creating this organization in Okta.":
      return (
        <Alert type="error" title="Unexpected error">
          An unexpected error occurred. Please resubmit this form, or contact{" "}
          <a href="mailto:support@simplereport.gov">support@simplereport.gov</a>{" "}
          if the problem persists.
        </Alert>
      );
    default:
      return (
        <Alert type="error" title="Submission error">
          {error} Please contact your organization administrator or{" "}
          <a href="mailto:support@simplereport.gov">support@simplereport.gov</a>{" "}
          for help.
        </Alert>
      );
  }
};
