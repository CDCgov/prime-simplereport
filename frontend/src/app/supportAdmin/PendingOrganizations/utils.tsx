import * as yup from "yup";

import { phoneNumberIsValid } from "../../patients/personSchema";

export interface PendingOrganizationFormValues {
  name: string;
  adminFirstName: string | undefined;
  adminLastName: string | undefined;
  adminEmail: string | undefined;
  adminPhone: string | undefined;
}

export interface EditOrgMutationResponse {
  data: {
    editPendingOrganization: string;
  };
}

const phoneNumberIsValidOrEmpty = (input: any) =>
  input === "" || phoneNumberIsValid(input);

export const pendingOrganizationSchema: yup.SchemaOf<PendingOrganizationFormValues> =
  yup.object().shape({
    name: yup.string().required("Organization name is required"),
    adminFirstName: yup.string(),
    adminLastName: yup.string(),
    adminEmail: yup.string().email("A valid email address is required"),
    adminPhone: yup
      .mixed()
      .test("", "A valid phone number is required", phoneNumberIsValidOrEmpty),
  });
