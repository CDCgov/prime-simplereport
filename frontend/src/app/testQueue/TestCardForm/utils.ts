import { SymptomDefinitionMap } from "../../../patientApp/timeOfTest/constants";

export const parseSymptoms = (
  symptomsJsonString: string | null | undefined,
  symptomSelectionMap: SymptomDefinitionMap[]
) => {
  const symptoms: Record<string, boolean> = {};

  if (symptomsJsonString) {
    const parsedSymptoms: { [key: string]: string | boolean } =
      JSON.parse(symptomsJsonString);

    symptomSelectionMap.forEach((opt) => {
      const val = opt?.value;
      if (typeof parsedSymptoms[val] === "string") {
        symptoms[val] = parsedSymptoms[val] === "true";
      } else {
        symptoms[val] = parsedSymptoms[val] as boolean;
      }
    });
  } else {
    symptomSelectionMap.forEach((opt) => {
      symptoms[opt?.value] = false;
    });
  }
  return symptoms;
};
