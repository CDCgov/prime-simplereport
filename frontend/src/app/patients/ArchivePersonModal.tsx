import React from "react";
import { gql, useMutation } from "@apollo/client";
import Modal from "react-modal";

import Button from "../commonComponents/Button/Button";
import { displayFullName } from "../utils";
import { showSuccess } from "../utils/srToast";
import "./ArchivePersonModal.scss";

import { Patient } from "./ManagePatients";

const ARCHIVE_PERSON_RECORD = gql`
  mutation ArchivePerson($id: ID!, $deleted: Boolean!) {
    setPatientIsDeleted(id: $id, deleted: $deleted) {
      internalId
    }
  }
`;

interface Props {
  person: Patient;
  closeModal: () => void;
}
interface ArchivePersonResponse {
  internalId: string;
}
interface ArchivePersonParams {
  id: string;
  deleted: boolean;
}
const ArchivePersonModal = ({ person, closeModal }: Props) => {
  const [archivePersonRecord] = useMutation<
    ArchivePersonResponse,
    ArchivePersonParams
  >(ARCHIVE_PERSON_RECORD);
  const yesArchivePerson = () => {
    archivePersonRecord({
      variables: {
        id: person.internalId,
        deleted: true,
      },
    })
      .then(() => {
        showSuccess("", "Record archived");
      })
      .finally(() => {
        closeModal();
      });
  };

  return (
    <Modal
      isOpen={true}
      className="sr-archive-person-modal-content"
      overlayClassName="sr-archive-person-modal-overlay"
      contentLabel="Archive record"
      onRequestClose={closeModal}
    >
      <p>
        Are you sure you want to archive the record for{" "}
        <b>
          {displayFullName(
            person.firstName,
            person.middleName,
            person.lastName
          )}
        </b>
        ?
      </p>
      <div className="sr-archive-person-buttons">
        <Button variant="unstyled" label="No, go back" onClick={closeModal} />
        <Button label="Yes, I'm sure" onClick={yesArchivePerson} />
      </div>
    </Modal>
  );
};

export default ArchivePersonModal;
