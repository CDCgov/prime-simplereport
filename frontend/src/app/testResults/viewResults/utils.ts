import { Location } from "react-router-dom";

import { getParameterFromUrl } from "../../utils/url";

export function getFiltersFromUrl(url: Location) {
  const patientId = getParameterFromUrl("patientId", url);
  const startDate = getParameterFromUrl("startDate", url);
  const endDate = getParameterFromUrl("endDate", url);
  const role = getParameterFromUrl("role", url);
  const disease = getParameterFromUrl("disease", url);
  const result = getParameterFromUrl("result", url);
  const filterFacilityId = getParameterFromUrl("filterFacilityId", url);

  return {
    ...(patientId && { patientId: patientId }),
    ...(startDate && { startDate: startDate }),
    ...(endDate && { endDate: endDate }),
    ...(disease && { disease: disease }),
    ...(result && { result: result }),
    ...(role && { role: role }),
    ...(filterFacilityId && { filterFacilityId: filterFacilityId }),
  };
}
