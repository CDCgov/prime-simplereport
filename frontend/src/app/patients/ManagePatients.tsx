import { gql, useQuery } from "@apollo/client";
import React, { useEffect, useState } from "react";
import moment from "moment";
import classnames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSlidersH } from "@fortawesome/free-solid-svg-icons";
import { useNavigate } from "react-router-dom";

import { displayFullName } from "../utils";
import { PATIENT_TERM, PATIENT_TERM_PLURAL_CAP } from "../../config/constants";
import { daysSince } from "../utils/date";
import { capitalizeText } from "../utils/text";
import { LinkWithQuery } from "../commonComponents/LinkWithQuery";
import { ActionsMenu } from "../commonComponents/ActionsMenu";
import {
  InjectedQueryWrapperProps,
  QueryWrapper,
} from "../commonComponents/QueryWrapper";
import Pagination from "../commonComponents/Pagination";
import { useDebounce } from "../testQueue/addToQueue/useDebounce";
import { SEARCH_DEBOUNCE_TIME } from "../testQueue/constants";
import Button from "../commonComponents/Button/Button";
import SearchInput from "../testQueue/addToQueue/SearchInput";
import { TestResult as TestResultType } from "../../app/testQueue/QueueItem";

import PatientUpload from "./PatientUpload";
import ArchivePersonModal from "./ArchivePersonModal";

import "./ManagePatients.scss";

export const patientsCountQuery = gql`
  query GetPatientsCountByFacility(
    $facilityId: ID!
    $showDeleted: Boolean!
    $namePrefixMatch: String
  ) {
    patientsCount(
      facilityId: $facilityId
      showDeleted: $showDeleted
      namePrefixMatch: $namePrefixMatch
    )
  }
`;

export const patientQuery = gql`
  query GetPatientsByFacility(
    $facilityId: ID!
    $pageNumber: Int!
    $pageSize: Int!
    $showDeleted: Boolean
    $namePrefixMatch: String
  ) {
    patients(
      facilityId: $facilityId
      pageNumber: $pageNumber
      pageSize: $pageSize
      showDeleted: $showDeleted
      namePrefixMatch: $namePrefixMatch
    ) {
      internalId
      firstName
      lastName
      middleName
      birthDate
      isDeleted
      role
      lastTest {
        dateAdded
      }
    }
  }
`;

export interface Patient {
  internalId: string;
  firstName: string;
  lastName: string;
  middleName: string;
  birthDate: string;
  isDeleted: boolean;
  role: string;
  lastTest: {
    dateAdded: string;
    result: TestResultType;
    dateTested: string;
    deviceTypeModel: string;
    deviceTypeName: string;
    facilityName: string;
  };
}

interface Props {
  activeFacilityId: string;
  canEditUser: boolean;
  canDeleteUser: boolean;
  currentPage: number;
  entriesPerPage: number;
  totalEntries?: number;
  showDeleted?: boolean;
  data?: { patients: Patient[] };
  refetch: () => null;
  setNamePrefixMatch: (namePrefixMatch: string | null) => void;
  isAdmin: boolean;
}

export const DetachedManagePatients = ({
  canEditUser,
  data,
  currentPage,
  entriesPerPage,
  totalEntries,
  refetch,
  setNamePrefixMatch,
  isAdmin,
  activeFacilityId,
}: Props) => {
  const [archivePerson, setArchivePerson] = useState<Patient | null>(null);
  const navigate = useNavigate();

  const [queryString, debounced, setDebounced] = useDebounce<string | null>(
    null,
    {
      debounceTime: SEARCH_DEBOUNCE_TIME,
    }
  );

  useEffect(() => {
    if (queryString && queryString.length > 1) {
      setNamePrefixMatch(queryString);
    } else if (!queryString) {
      setNamePrefixMatch(null);
    }
    navigate({
      search: `?facility=${activeFacilityId}`,
    });
  }, [queryString, setNamePrefixMatch, navigate, activeFacilityId]);

  if (archivePerson) {
    return (
      <ArchivePersonModal
        person={archivePerson}
        closeModal={() => {
          setArchivePerson(null);
          refetch();
        }}
      />
    );
  }

  const patientRows = (patients: Patient[]) => {
    if (patients.length === 0) {
      return (
        <tr>
          <td colSpan={5}>No results</td>
        </tr>
      );
    }

    return patients.map((patient: Patient) => {
      let fullName = displayFullName(
        patient.firstName,
        patient.middleName,
        patient.lastName
      );

      const editUserLink =
        canEditUser && !patient.isDeleted ? (
          <LinkWithQuery
            to={`/patient/${patient.internalId}`}
            className="sr-patient-edit-link"
          >
            {fullName}
          </LinkWithQuery>
        ) : (
          <span>{fullName}</span>
        );

      return (
        <tr
          key={patient.internalId}
          className={classnames(
            "sr-patient-row",
            patient.isDeleted && "sr-patient-row--removed"
          )}
        >
          <th scope="row">{editUserLink}</th>
          <td>{moment(patient.birthDate).format("MM/DD/yyyy")}</td>
          <td>{capitalizeText(patient.role)}</td>
          <td>
            {patient.lastTest
              ? `${daysSince(moment(patient.lastTest.dateAdded))}`
              : "N/A"}
          </td>
          <td>
            {canEditUser && !patient.isDeleted && (
              <ActionsMenu
                items={[
                  {
                    name: "Archive record",
                    action: () => setArchivePerson(patient),
                  },
                ]}
              />
            )}
          </td>
        </tr>
      );
    });
  };

  return (
    <main className="prime-home">
      <div className="grid-container">
        <div className="grid-row">
          <div className="prime-container card-container">
            <div className="usa-card__header">
              <h2>
                {PATIENT_TERM_PLURAL_CAP}
                <span className="sr-showing-patients-on-page">
                  {totalEntries === undefined ? (
                    "Loading..."
                  ) : (
                    <>
                      Showing {(currentPage - 1) * entriesPerPage + 1}-
                      {Math.min(entriesPerPage * currentPage, totalEntries)} of{" "}
                      {totalEntries}
                    </>
                  )}
                </span>
              </h2>
              <div>
                <Button
                  className="sr-active-button"
                  icon={faSlidersH}
                  onClick={() => {
                    setNamePrefixMatch(null);
                    setDebounced(null);
                  }}
                >
                  Clear filters
                </Button>
                {canEditUser ? (
                  <LinkWithQuery
                    className="usa-button usa-button--primary"
                    to={`/add-patient`}
                    id="add-patient-button"
                  >
                    <FontAwesomeIcon icon="plus" />
                    {` Add ${PATIENT_TERM}`}
                  </LinkWithQuery>
                ) : null}
              </div>
            </div>
            <div className="display-flex flex-row bg-base-lightest padding-x-3 padding-y-2">
              <SearchInput
                label="Person"
                onInputChange={(e) => {
                  setDebounced(e.target.value);
                }}
                onSearchClick={(e) => {
                  e.preventDefault();
                  setNamePrefixMatch(debounced);
                }}
                queryString={debounced || ""}
                className="display-inline-block"
                placeholder=""
                focusOnMount
              />
            </div>
            <div className="usa-card__body sr-patient-list">
              <table className="usa-table usa-table--borderless width-full">
                <thead>
                  <tr>
                    <th scope="col">Name</th>
                    <th scope="col">Date of birth</th>
                    <th scope="col">Type</th>
                    <th scope="col">Days since last test</th>
                    <th scope="col">Actions</th>
                  </tr>
                </thead>
                <tbody aria-live="polite">
                  {data ? (
                    patientRows(data.patients)
                  ) : (
                    <tr>
                      <td colSpan={5}>Loading...</td>
                    </tr>
                  )}
                </tbody>
              </table>
            </div>
            {data?.patients && data.patients.length > 0 && (
              <div className="usa-card__footer">
                {totalEntries && (
                  <Pagination
                    baseRoute="/patients"
                    currentPage={currentPage}
                    entriesPerPage={entriesPerPage}
                    totalEntries={totalEntries}
                  />
                )}
              </div>
            )}
          </div>
          {isAdmin && <PatientUpload onSuccess={refetch} />}
        </div>
      </div>
    </main>
  );
};

type InjectedContainerProps =
  | "pageCount"
  | "entriesPerPage"
  | "totalEntries"
  | "setNamePrefixMatch";

const ManagePatients = (
  props: Omit<Props, InjectedQueryWrapperProps | InjectedContainerProps>
) => {
  const [namePrefixMatch, setNamePrefixMatch] = useState<string | null>(null);

  const { data: totalPatients, error, refetch: refetchCount } = useQuery(
    patientsCountQuery,
    {
      variables: {
        facilityId: props.activeFacilityId,
        showDeleted: false,
        namePrefixMatch,
      },
      fetchPolicy: "no-cache",
    }
  );

  if (props.activeFacilityId.length < 1) {
    return <div>"No facility selected"</div>;
  }

  if (error) {
    throw error;
  }

  const totalEntries = totalPatients?.patientsCount;
  const entriesPerPage = 20;
  const pageNumber = props.currentPage || 1;
  return (
    <QueryWrapper<Props>
      query={patientQuery}
      queryOptions={{
        variables: {
          facilityId: props.activeFacilityId,
          pageNumber: pageNumber - 1,
          pageSize: entriesPerPage,
          showDeleted: props.showDeleted || false,
          namePrefixMatch,
        },
      }}
      onRefetch={refetchCount}
      Component={DetachedManagePatients}
      displayLoadingIndicator={false}
      componentProps={{
        ...props,
        totalEntries,
        currentPage: pageNumber,
        entriesPerPage,
        setNamePrefixMatch,
      }}
    />
  );
};

export default ManagePatients;
