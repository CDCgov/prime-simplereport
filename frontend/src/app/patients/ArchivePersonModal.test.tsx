import { render, screen, waitFor } from "@testing-library/react";
import ReactDOM from "react-dom";
import { MockedProvider } from "@apollo/client/testing";
import userEvent from "@testing-library/user-event";

import { ArchivePersonDocument } from "../../generated/graphql";
import * as srToast from "../utils/srToast";

import { Patient } from "./ManagePatients";
import ArchivePersonModal from "./ArchivePersonModal";

const mockPerson: Patient = {
  internalId: "7b968f75-e2fb-43a5-ae8b-c7e0f4873d3a",
  firstName: "Guy",
  lastName: "Abramcik",
  middleName: "Christine Michael",
  birthDate: "1992-11-26",
  isDeleted: false,
  role: "STAFF",
  lastTest: {
    dateAdded: "2022-01-06 13:56:20.255",
    result: "",
    dateTested: "2022-01-07 13:56:20.255",
    deviceTypeModel: "",
    deviceTypeName: "",
    facilityName: "",
  },
};

const archivePersonMutation = {
  request: {
    query: ArchivePersonDocument,
    variables: {
      id: "7b968f75-e2fb-43a5-ae8b-c7e0f4873d3a",
      deleted: true,
    },
  },
  result: {
    data: {
      setPatientIsDeleted: {
        internalId: "c912d4d4-cbe6-4d80-9d24-c14ba1f7f180",
        deleted: true,
      },
    },
  },
};

const setup = () => {
  ReactDOM.createPortal = jest.fn((element, _node) => {
    return element;
  }) as any;
  return render(
    <MockedProvider mocks={[archivePersonMutation]}>
      <ArchivePersonModal person={mockPerson} closeModal={() => {}} />
    </MockedProvider>
  );
};

describe("ArchivePersonModal", () => {
  it("contains the defaults", async () => {
    let view = setup();
    expect(
      await screen.findByText("Abramcik, Guy Christine Michael")
    ).toBeInTheDocument();
    expect(view).toMatchSnapshot();
  });

  it("shows a success message when archiving", async () => {
    let alertSpy: jest.SpyInstance = jest.spyOn(srToast, "showSuccess");
    setup();
    userEvent.click(screen.getByText("Yes, I'm sure"));
    await waitFor(() => {
      expect(alertSpy).toHaveBeenCalledWith("", "Record archived");
    });
  });
});
