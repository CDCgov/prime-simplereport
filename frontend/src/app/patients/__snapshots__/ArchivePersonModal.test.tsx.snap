// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchivePersonModal contains the defaults 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    aria-hidden="true"
    class="ReactModal__Body--open"
  >
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div>
      <div
        class="ReactModal__Overlay sr-archive-person-modal-overlay"
      >
        <div
          aria-label="Archive record"
          aria-modal="true"
          class="ReactModal__Content sr-archive-person-modal-content"
          role="dialog"
          tabindex="-1"
        >
          <p>
            Are you sure you want to archive the record for
             
            <b>
              Abramcik, Guy Christine Michael
            </b>
            ?
          </p>
          <div
            class="sr-archive-person-buttons"
          >
            <button
              class="usa-button usa-button--unstyled"
              type="button"
            >
              No, go back
            </button>
            <button
              class="usa-button"
              type="button"
            >
              Yes, I'm sure
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      data-react-modal-body-trap=""
      style="position: absolute; opacity: 0;"
      tabindex="0"
    />
    <div
      class="ReactModalPortal"
    />
  </body>,
  "container": <div>
    <div
      class="ReactModal__Overlay sr-archive-person-modal-overlay"
    >
      <div
        aria-label="Archive record"
        aria-modal="true"
        class="ReactModal__Content sr-archive-person-modal-content"
        role="dialog"
        tabindex="-1"
      >
        <p>
          Are you sure you want to archive the record for
           
          <b>
            Abramcik, Guy Christine Michael
          </b>
          ?
        </p>
        <div
          class="sr-archive-person-buttons"
        >
          <button
            class="usa-button usa-button--unstyled"
            type="button"
          >
            No, go back
          </button>
          <button
            class="usa-button"
            type="button"
          >
            Yes, I'm sure
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
